File: Algorithms/Sorting/merge_sort.c
#include <stdio.h>
#include <stdlib.h>

void
merge (int* seq, int low, int mid, int high)
{
  int* left = malloc (sizeof (int) * (mid - low + 2));
  int* right = malloc (sizeof (int) * (high - mid + 1));

  int i, j;

  for (i = 0; i < (mid - low + 1); ++i)
    left[i] = seq[low + i];
  left[i] = 10;

  for (i = 0; i < (high - mid); ++i)
    right[i] = seq[mid + 1 + i];
  right[i] = 10;

  i = 0;
  j = 0;
  for (int k = low; k <= high; ++k)
    {
      if (left[i] <= right[j])
        seq[k] = left[i++];
      else
        seq[k] = right[j++];
    }

  free (left);
  free (right);
}

void
merge_sort (int* seq, int low, int high)
{
  if (low < high)
    {
      int mid = (low + high) / 2;
      merge_sort (seq, low, mid);
      merge_sort (seq, mid + 1, high);
      merge (seq, low, mid, high);
    }
}


int
main ()
{
  int seq[] = {3, 5, 2, 4, 1};
  int n = 5;

  merge_sort (seq, 0, n - 1);

  for (int i = 0; i < n; ++i)
    printf ("%d ", seq[i]);
  printf ("\n");
}

File: Algorithms/Sorting/merge_sort.rs
fn merge(seq: &mut [u8; 5], low: usize, mid: usize, high: usize) {
    let mut left: Vec<u8> = seq[low..mid + 1].to_vec();
    let mut right: Vec<u8> = seq[mid + 1..high + 1].to_vec();

    left.push(10);
    right.push(10);
    
    let mut i: usize = 0;
    let mut j: usize = 0;
    for k in low..high + 1 {
        if left[i] <= right[j] {
            seq[k] = left[i];
            i += 1;
        } else {
            seq[k] = right[j];
            j += 1;
        }
    }
}

fn merge_sort(seq: &mut [u8; 5], low: usize, high: usize) {
    if low < high {
        let mid: usize = (low + high) / 2;
        merge_sort (seq, low, mid);
        merge_sort (seq, mid + 1, high);
        merge (seq, low, mid, high);
    }
}

fn main() {
    let mut seq: [u8; 5] = [3, 5, 2, 4, 1];

    merge_sort(&mut seq, 0, 4);

    for i in &seq {
        print!("{} ", i);
    }
    print!("\n");
}
